### ISSUE TITLE
Why are the inputs and labels to the model both the same while training? 

### ISSUE THREAD

**Issue by desaibhargav** [cdpierse/script_buddy_v2#1]
I'm currently working on a project where I need to fine-tune GPT2 and I came across your work which btw I think is wonderful. But I had difficulty understanding why the inputs and the labels passed in the training loop are the same. Won't the model learn something equivalent to an identity function or a copying mechanism? 
I mean, clearly it works but if you could just lay out the concept for me that would be great!

**Comment by cdpierse** [comment]
Hey @algoromeo thanks. Yes it can seem a bit confusing at first. It's because we are using GPT2 with a language modelling head. If you check out Hugging Faces's documentation [here ](https://huggingface.co/transformers/model_doc/gpt2.html#gpt2lmheadmodel) in the parameters section it describes how the labels are shifted inside the model itself. 

The reason this is done is because it's CLM (causal language model), while there are no steps in Transformers like RNN's it might help to imagine that at each step or position in the inputs the model can only see all the inputs up to that point and what it is trying to predict is the next token thus why the labels are shifted. Inputs *0-n* are used to predict output *n+1*. This is done for every position in the input thus why we need to feed in the input_id's as labels. 

So while the input_ids are the labels the model is learning weights that predict one token at a time in an auto regressive manner. Does this help?

**Comment by desaibhargav** [comment]
> Hey @algoromeo thanks. Yes it can seem a bit confusing at first. It's because we are using GPT2 with a language modelling head. If you check out Hugging Faces's documentation [here ](https://huggingface.co/transformers/model_doc/gpt2.html#gpt2lmheadmodel) in the parameters section it describes how the labels are shifted inside the model itself.
> 
> The reason this is done is because it's CLM (causal language model), while there are no steps in Transformers like RNN's it might help to imagine that at each step or position in the inputs the model can only see all the inputs up to that point and what it is trying to predict is the next token thus why the labels are shifted. Inputs _0-n_ are used to predict output _n+1_. This is done for every position in the input thus why we need to feed in the input_id's as labels.
> 
> So while the input_ids are the labels the model is learning weights that predict one token at a time in an auto regressive manner. Does this help?

That makes perfect sense! Thank you!
